build/commands/lib/test.js

```diff
@@ -115,6 +115,80 @@ const buildTests = async (
   await util.buildTargets()
 }
 
+const getResultsFileInfo = (file) => {
+  if (!fs.existsSync(file)) {
+    console.log(`Error: file ${file} doesn't exist`)
+    return null
+  }
+  try {
+    const data = fs.readFileSync(file, 'utf8')
+    const testsuitesPattern =
+      '<testsuites name="AllTests"'
+      + ' tests="(?<tests>[0-9]+)" failures="(?<failures>[0-9]+)"'
+      + ' disabled="(?<disabled>[0-9]+)" errors="(?<errors>[0-9]+)"'
+      + ' time="(?<time>[0-9]+(.[0-9]+)?)"'
+      + ' timestamp="(?<timestamp>[0-9\-:T]+)">'
+    const testsuites = data.match(testsuitesPattern)
+    if (!testsuites) {
+      console.log(`Could not find matching testsuites element in ${file}`)
+      return null
+    }
+    // Strip xml declaration and the `testsuites` tag (doing this with brute
+    // force since bringing in a new xml parser is too heavy for this.)
+    const newData = data
+      .replace('<?xml version="1.0" encoding="UTF-8"?>', '')
+      .replace(testsuites[0], '')
+      .replace('</testsuites>', '')
+    return {
+      data: newData.replace(/^\s*\n/gm, ''),
+      ...testsuites.groups,
+    }
+  } catch (err) {
+    console.log(`Error reading results file ${file}: ` + err)
+    return null
+  }
+}
+
+// Merge test results from `from` file to `into` file
+const mergeTestOutput = (suite, from, into) => {
+  assert(
+    from !== into,
+    `Attempted to merge test output into itself. File: ${from}`,
+  )
+  if (!fs.existsSync(into)) {
+    fs.copyFileSync(from, into)
+    return true
+  }
+  fromInfo = getResultsFileInfo(from)
+  if (!fromInfo) {
+    console.log(`Error getting results file info for ${from}.`)
+    return false
+  }
+  intoInfo = getResultsFileInfo(into)
+  if (!intoInfo) {
+    console.log(`Error getting results file info for ${into}.`)
+    return false
+  }
+  let newData = '<?xml version="1.0" encoding="UTF-8"?>\n'
+  newData +=
+    '<testsuites name="AllTests"'
+    + ` tests="${Number(intoInfo.tests) + Number(fromInfo.tests)}"`
+    + ` failures="${Number(intoInfo.failures) + Number(fromInfo.failures)}"`
+    + ` disabled="${Number(intoInfo.disabled) + Number(fromInfo.disabled)}"`
+    + ` errors="${Number(intoInfo.errors) + Number(fromInfo.errors)}"`
+    + ` time="${Number(intoInfo.time) + Number(fromInfo.time)}"`
+    + ` timestamp="${intoInfo.timestamp}">\n`
+  newData += intoInfo.data + fromInfo.data + '</testsuites>'
+
+  try {
+    fs.writeFileSync(into, newData)
+  } catch (err) {
+    console.log(`Error merging ${from} into ${into} file: ` + err)
+    return false
+  }
+  return true
```

Comment:
I see too much moving parts here. I suggest generating a separate xml for each test suite with a prefix passed via a command line. CI should be modified to aggregate test results by mask only from files with this prefix (either a build number or a timestamp).

All this xml fiddling is a bit too much.

Reply:
@wknapik thoughts?

Reply:
Merging them seems overly complex to me and I think we'd like to control which tests get run based on dependencies in gn. I think it would be better to output a known file that has the list of all of the individual test outputs and CI can use that to display the results

Reply:
so for instance
```test-results
brave_unit_tests.xml
brave_components_unit_test.xml
...
```
and CI reads that file to get all the tests that ran

Reply:
I think that's also better because the results match up with the individual test suites instead of being grouped together

Reply:
@wknapik looks like an additional `.xml` is being added when parsing result files: 
```
Parsing src/brave_unit_tests.xml.xml,src/brave_installer_unittests.xml.xml
```

Reply:
@wknapik:
```
reports = readFile("src/${testType}.txt").tokenize().collect { "src/${it}.xml" }.join(',')
```
adds `.xml` to `it`.

The `.txt` file contains names with (xml) extension already:
```
        // Add xml output filename of each test suite into the results file
        if (!appendToResultsFile(allResultsFilePath, `${testSuite}.xml`)) {
          return false
        }
```